Re:Tổng hợp các câu hỏi? Thanh Pham Minh wrote: _Tieu_ Doan_Du wrote: Đó là tư duy thông thường, nếu thế thì chả cần phải nói khái niệm biến phải được gán trước break làm gì, vì theo lẽ thường thì sau break có thực thi đâu. OK, mình sẽ nói kỹ hơn một xíu. Tạm thời chúng ta có 3 lệnh cần quan tâm: lệnh return, lệnh gán, lệnh sử dụng (biến). Giờ xem như chưa có lệnh return: trong function chỉ có lệnh gán và lệnh sử dụng. Chúng ta sẽ đưa lệnh return vào sau để thấy được sự logic thông thường. Trường hợp 1: không thể chắc chắn lệnh gán xảy ra, lúc này ở lệnh sử dụng sẽ bị lỗi Not Defined Before Use. Giờ dù ta có đưa return vào chỗ nào đi nữa cũng không thay đổi được vấn đề (lệnh gán vẫn không thể chắc chắn xảy ra hay không => ở lệnh sử dụng vẫn lỗi). Trường hợp 2: lệnh gán chắc chắn xảy ra, lúc này ở lệnh sử dụng không có lỗi. 2a) Giờ nếu ta đưa return vào sau cả hai lệnh trên thì vẫn không có lỗi. 2b) Nếu ta đưa return vào giữa 2 lệnh trên. Chia ra: i) Nếu return xảy ra => lệnh sử dụng (nằm sau lệnh return) không thể xảy ra => không thể có lỗi (vẫn giống lúc không có return). ii)) Nếu return không xảy ra: trước lệnh sử dụng vẫn chắc chắn có lệnh gán (nếu không chắc chắn có lệnh gán thì rơi vào trường hợp 1 rồi nhé) => vẫn không thể có lỗi. iii) Return không chắc có xảy ra không: dù thế nào cũng là 1 trong 2 cái (i, ii) đã nói ở trên. 2c) Nếu ta đưa lệnh return vào trước 2 lệnh trên. Chia ra: i) Nếu return xảy ra => lệnh sử dụng (nằm sau lệnh return) không thể  xảy ra => không thể có lỗi (vẫn giống lúc không có return). ii)) Nếu return không xảy ra: trước lệnh sử dụng vẫn chắc chắn có lệnh  gán (nếu không chắc chắn có lệnh gán thì rơi vào trường hợp 1 rồi nhé)  => vẫn không thể có lỗi. iii) Return không chắc có xảy ra không: dù thế nào cũng là 1 trong 2 cái (i, ii) đã nói ở trên. Ok, rất là dài dòng, nhưng nếu bạn suy nghĩ được một cách thông thường thì không khó hiểu lắm đâu. Giờ đơn giản nhất là bạn hãy lấy một ví dụ để thấy rằng khi có và không có return thì kết quả (lỗi hay không lỗi Not Define...) thay đổi (theo ý mà bạn đầu topic đã nói). Nếu tìm được, tức là mình sai (& vice versa). Edit: à, mình hiểu rồi, thực ra bạn nhầm ở chỗ này. Sau break các lệnh nằm ngoài loop (vẫn trong function) CÓ thực thi bạn nhé => biến local của function vẫn còn. Còn sau return thì thoát khỏi function mất rồi, biến global thì mặc định được coi là đã define (theo đặc tả). Cái này chỉ cần anh Hùng nói một câu thì khỏi mất công suy diễn, mong anh Hùng giải đáp, sắp đến deadline rồi ạ.string trong Assign ? 2 lỗi đó cùng định dạng mà bạn: Type Mismatch In Statement: assign(test,"'string test'")